File: Notes/4.RepeatDsForPython/DataStructuresforCodingInterviewsinPython/1.lists/Challenge-2/go/solution.go
package main

import (
  "fmt"
  "reflect"
)

func main() {
  arr1 := []int{1, 3, 4, 5}
  arr2 := []int{2, 6, 7, 8}

  var result []int = merge(arr1, arr2)
  expected := []int{1, 2, 3, 4, 5, 6, 7, 8}

  if reflect.DeepEqual(result, expected) {
    fmt.Println("Success")
  } else {
    fmt.Println("Failure")
  }
}

func merge(arr1 []int, arr2 []int) []int {
    index1 := 0
    index2 := 0
    index_res := 0
    var res []int

    for i := 0; i < len(arr1) + len(arr2); i++ {
        res = append(res, 0)
    }

    for (index1 < len(arr1)) && (index2 < len(arr2)) {
      if arr1[index1] < arr2[index2] {
        res[index_res] = arr1[index1]
        index1 ++
        index_res ++
      } else {
        res[index_res] = arr2[index2]
        index2 ++
        index_res ++
      }
    }

    for index1 < len(arr1) {
      res[index_res] = arr1[index1]
      index1 ++
      index_res ++
    }

    for index2 < len(arr2) {
      res[index_res] = arr2[index2]
      index2 ++
      index_res ++
    }

    return res
}

File: Notes/4.RepeatDsForPython/DataStructuresforCodingInterviewsinPython/1.lists/Challenge-2/js/solution.js
let merge = (arr1, arr2) => {
  let index1 = 0;
  let index2 = 0;
  let index_res = 0;
  let res = [];

  for (let i = 0; i < arr1.length + arr2.length; i++) {
    res.push(0);
  }

  while ((index1 < arr1.length) && (index2 < arr2.length)) {
    if (arr1[index1] < arr2[index2]) {
      res[index_res] = arr1[index1];
      index1 ++;
      index_res ++;
    } else {
      res[index_res] = arr2[index2];
      index2 ++;
      index_res ++;
    }
  }

  while(index1 < arr1.length) {
    res[index_res] = arr1[index1];
    index1 ++;
    index_res ++;
  }

  while(index2 < arr2.length) {
    res[index_res] = arr2[index2];
    index2 ++;
    index_res ++;
  }

  return res;
}

result =  merge([1, 3, 4, 5],[2, 6, 7, 8]);
expected = [1, 2, 3, 4, 5, 6, 7, 8];

let equalArrays = (a,b) => {
  if (a.length != b.length) return false;
  for(var i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) return false;
  }
  return true;
}

if (equalArrays(result, expected)) {
  console.log("Success");
} else {
  console.log("Failure");
}

